pkg/ibm_plugin/server/plugin.go

```diff
@@ -326,136 +327,207 @@ func (s *IBMPluginServer) AddPermitListRules(ctx context.Context, req *paraglide
-		clients, err := s.getAllClientsForVPCs(cloudClient, rInfo.ResourceGroup, false)
+	for _, paragliderRule := range req.Rules {
+		// translate paraglider rule to IBM rules to compare hash values with current rules.
+		// multiple ibm rules can be returned due to multiple possible targets
+		ibmRules, err := sdk.ParagliderToIBMRule(requestSGID, paragliderRule)
 		if err != nil {
 			utils.Log.Printf("Failed to get remote vpc: %v.\n", err)
 			return nil, err
 		}
-		remoteVPC := ""
-		for vpcID, client := range clients {
-			if isRemoteInVPC, _ := client.IsRemoteInVPC(vpcID, ibmRule.Remote); isRemoteInVPC {
-				remoteVPC = vpcID
-				break
-			}
+
+		// peeringCloudInfos contains cloud info associated with the specified rule's targets
+		peeringCloudInfos, err := utils.GetPermitListRulePeeringCloudInfo(paragliderRule, addressSpaceMappings.AddressSpaceMappings)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get peering cloud infos: %w", err)
 		}
-		// if the remote resides inside a paraglider VPC that isn't the request resource's VPC, connect them
-		if remoteVPC != "" && remoteVPC != *requestVPCData.ID {
-			utils.Log.Printf("The following rule's remote is targeting a different IBM VPC\nRule: %+v\nVPC:%+v", ibmRule, remoteVPC)
-			// fetch or create transit gateway
-			if len(gwID) == 0 { // lookup optimization, use the already fetched gateway ID if possible
-				gwID, err = cloudClient.GetOrCreateTransitGateway(region)
+		// connect clouds if needed
+		for i, peeringCloudInfo := range peeringCloudInfos {
+			if peeringCloudInfo == nil {
+				continue
+			}
+			if peeringCloudInfo.Cloud != utils.IBM {
+				ruleTargetAddress := ibmRules[i].Remote
+				// Create VPN connections
+				connectCloudsReq := &paragliderpb.ConnectCloudsRequest{
+					CloudA:             utils.IBM,
+					CloudANamespace:    req.Namespace,
+					CloudB:             peeringCloudInfo.Cloud,
+					CloudBNamespace:    peeringCloudInfo.Namespace,
+					AddressSpaceCloudA: vpcAddressSpaces,
+					AddressSpaceCloudB: []string{ruleTargetAddress},
+				}
+				if len(ruleTargetAddress) == 0 {
+					return nil, fmt.Errorf("Missing remote address for rule %+v", ibmRules[i])
+				}
+				_, err := controllerClient.ConnectClouds(ctx, connectCloudsReq)
+				if err != nil {
+					return nil, fmt.Errorf("unable to connect clouds : %w", err)
+				}
+			} else {
+				// if rule targets a VPC on IBM, connect them via a transit gateway
+				err = s.connectToTransitGatewayIfNeeded(cloudClient, ibmRules[i], gwID, rInfo.ResourceGroup, *requestVPCData.CRN, region)
 				if err != nil {
 					return nil, err
 				}
 			}
-			// connect the VPC of the request's resource to the transit gateway.
-			// the `remoteVPC` should be connected by a separate symmetric request (e.g. to allow inbound traffic to remote).
-			err = cloudClient.ConnectVPC(gwID, *requestVPCData.CRN)
+		}
+
+		// add unique rules to security group
+		for _, ibmRule := range ibmRules {
```

Comment:
Hey @smcclure20, I'm strongly against skipping this step.
- [IBM's security group rules](https://pkg.go.dev/github.com/IBM/vpc-go-sdk@v0.51.0/vpcv1#SecurityGroupRulePrototype) don't have a `name` field, nor are they taggable, so your example creates identical entities from the cloud's perspective. Your example input (two identical rules with different tags) will not lead to the creation of 2 rules, as the second insertion attempt will be [refused](https://github.com/paraglider-project/paraglider/blob/699907c70ec31499c776d12118b1526d160c5088/pkg/ibm_plugin/server/plugin.go#L416). 
- Removing the unique property check, will also remove the idempotent property from `AddPermitListRules`, cluttering the security group with duplicates.  
To avoid this implication, the user will have to manually look up existing permit rules outside paraglider (because it doesn't support resource fetchers), before invoking this function via `glide`, shifting responsibilities to users and breaking their natural work flow.

Reply:
@Cohen-J-Omer:
Correct me if I am wrong: If I add two rules with same intent of allowing ping between host A and host B, but, with different names. The second rule will fail with `Rule %+v already exists for security group ID %v.`

@smcclure20 Is this behavior expected? since two identical rules are not allowed. If the above behavior is not expected, we can have two options : 

1) The second rule name would override the first rule name, and only one rule is added in the security group. 
2) The second rule is added, But Both rule names will point to same rule ID (underlying rule in the security group). We have to handle the deletion and modification correctly.

Reply:
I think it is preferred to fail in that case since the cloud can't support it and option 2 requires too much work on our end imo.

@Cohen-J-Omer : There is no idempotence property any more though, at least not when defining the identity of the rule as it properties. Rules, from the paraglider perspective, are identified by their names. And if an user tries to create a rule with the same name as one that exists, it should *update* that rule (or leave it the same if nothing has changed). However, if there are two rules with the same parameters but different names, this should fail for the reasons stated above. So I still don't see a need for deduplication on our end.

Reply:
OK, That makes sense. 
@smcclure20  I have created https://github.com/paraglider-project/paraglider/issues/389  and assigned myself based on the discussion, since this is a day 0 code, and not related to this PR :)